### PR TITLE
Crimson cmp xattr canceled operation

### PR DESCRIPTION
**Pull request by liu-chunmei**
## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/osd/pg_backend.cc [hunk 1]
```diff
@@ -930,37 +930,51 @@ PGBackend::get_attr_ierrorator::future<> PGBackend::get_xattrs(
...
 template<typename U, typename V>
-int do_cmp_xattr(int op, const U& lhs, const V& rhs)
+static PGBackend::cmp_xattr_errorator::future<>
+do_cmp_xattr(int op, const U& lhs, const V& rhs)
 {
+  bool result = 0;
   switch (op) {
   case CEPH_OSD_CMPXATTR_OP_EQ:
-    return lhs == rhs;
+    result = (lhs == rhs) ? 1 : 0;
```
Comment (athanatos) [inline src/crimson/osd/pg_backend.cc:939]:
```
result = (lhs == rhs);
```
should suffice.

File: src/crimson/osd/pg_backend.cc [hunk 2]
```diff
@@ -930,37 +930,51 @@ PGBackend::get_attr_ierrorator::future<> PGBackend::get_xattrs(
...
   default:
-    return -EINVAL;
+    return crimson::ct_error::invarg::make();
+  }
+  if (!result) {
+    return crimson::ct_error::ecanceled::make();
+  } else {
+    return PGBackend::cmp_xattr_errorator::now();
   }
 }
```
Comment (athanatos) [inline src/crimson/osd/pg_backend.cc:964]: Seems like making this return a future is unnecessarily complicated, why not do this in the caller?

File: src/crimson/osd/pg_backend.cc [hunk 3]
```diff
@@ -930,37 +930,51 @@ PGBackend::get_attr_ierrorator::future<> PGBackend::get_xattrs(
...
+do_xattr_cmp_u64(int op, uint64_t lhs, bufferlist&& rhs_xattr)
 {
   uint64_t rhs;
 
   if (rhs_xattr.length() > 0) {
     const char* first = rhs_xattr.c_str();
     if (auto [p, ec] = std::from_chars(first, first + rhs_xattr.length(), rhs);
 	ec != std::errc()) {
-      return -EINVAL;
+      return crimson::ct_error::invarg::make();
```
Comment (athanatos) [inline src/crimson/osd/pg_backend.cc:977]: Same here, I'd rather these returned error codes for clarity.

File: src/crimson/osd/pg_backend.cc [hunk 4]
```diff
@@ -969,57 +983,61 @@ static int do_xattr_cmp_u64(int op, uint64_t lhs, bufferlist& rhs_xattr)
...
+  object_stat_sum_t& delta_stats) const
+{
+  std::string name{"_"};
+  auto bp = osd_op.indata.cbegin();
+  bp.copy(osd_op.op.xattr.name_len, name);
+
+  logger().debug("cmpxattr on obj={} for attr={}", os.oi.soid, name);
+  return getxattr(os.oi.soid, std::move(name)).safe_then_interruptible(
+    [&delta_stats, &osd_op] (auto &&xattr) -> cmp_xattr_ierrorator::future<> {
+    return do_cmp_xattr_mode(osd_op, std::move(xattr)).safe_then_interruptible(
```
Comment (athanatos) [inline src/crimson/osd/pg_backend.cc:1034]: do_cmp_xattr_mode doesn't do any IO, right?  Can it simply return an error code?

File: src/crimson/os/seastore/seastore.cc [hunk 5]
```diff
@@ -704,7 +704,9 @@ SeaStore::_omap_get_value_ret SeaStore::_omap_get_value(
       return manager.omap_get_value(root, t, key
       ).si_then([](auto opt) -> _omap_get_value_ret {
         if (!opt) {
-          return crimson::ct_error::enodata::make();
+          bufferlist bl;
+          bl.append("");
+          return seastar::make_ready_future<ceph::bufferlist>(std::move(bl));
```
Comment (athanatos) [inline src/crimson/os/seastore/seastore.cc:709]: Are we sure this is the correct behavior?  BlueStore::getattr() returns -ENODATA if the attr isn't present.

Comment (athanatos) [reply src/crimson/os/seastore/seastore.cc:709]: Seems like we should handle ct_error::enodata in PGBackend::cmp_xattr.

Comment (liu-chunmei) [reply src/crimson/os/seastore/seastore.cc]: @athanatos updated in PGBackend::cmp_xattr. please review.

File: src/crimson/osd/pg_backend.cc [hunk 6]
```diff
@@ -1015,12 +1018,22 @@ PGBackend::cmp_xattr_ierrorator::future<> PGBackend::cmp_xattr(
     if (result == 0) {
       logger().info("cmp_xattr: comparison returned false");
       osd_op.rval = -ECANCELED;
+      return crimson::ct_error::ecanceled::make();
+    } else if (result == -EINVAL) {
+      return crimson::ct_error::invarg::make();
```
Comment (athanatos) [inline src/crimson/osd/pg_backend.cc:1023]: osd_op.rval = -EINVAL?

File: src/crimson/osd/pg_backend.cc [hunk 7]
```diff
@@ -1015,12 +1018,22 @@ PGBackend::cmp_xattr_ierrorator::future<> PGBackend::cmp_xattr(
...
+      return cmp_xattr_ierrorator::now();
     }
-    delta_stats.num_rd++;
-    delta_stats.num_rd_kb += shift_round_up(osd_op.op.xattr.value_len, 10);
-  });
+  }).handle_error_interruptible(
+    crimson::ct_error::enodata::handle([&delta_stats, &osd_op] ()
+      ->cmp_xattr_errorator::future<> {
+      delta_stats.num_rd++;
+      delta_stats.num_rd_kb += shift_round_up(osd_op.op.xattr.value_len, 10);
```
Comment (athanatos) [inline src/crimson/osd/pg_backend.cc:1032]: Since this is unconditional, why not do it once prior to the call to getattr?

Comment (liu-chunmei) [reply src/crimson/osd/pg_backend.cc:1032]: classic osd do it after calling getattr, seems there is other error code, which will not count it.

Comment (liu-chunmei) [reply src/crimson/osd/pg_backend.cc:1032]: such as ENOENT.

Comment (athanatos) [reply src/crimson/osd/pg_backend.cc:1032]: Hmm, ok.

**Review by athanatos — APPROVED** [review]

File: src/crimson/osd/pg_backend.cc (hunk 6 above)
Comment (athanatos) [reply src/crimson/osd/pg_backend.cc:1023]: Oh, still waiting on this one I think.  Otherwise LGTM

Comment (liu-chunmei) [reply src/crimson/osd/pg_backend.cc:1022]: remove osd_op.rval = -ECANCELED; it is not necessary here, error result will get from error_code,  osd_op_rval is used to save operation result.

Comment (liu-chunmei) [reply src/crimson/osd/pg_backend.cc:1022]: @athanatos any comments here?

Comment (athanatos) [reply src/crimson/osd/pg_backend.cc:1022]: ok, sounds good

File: src/crimson/osd/pg_backend.cc [hunk 8]
```diff
@@ -1012,15 +1015,24 @@ PGBackend::cmp_xattr_ierrorator::future<> PGBackend::cmp_xattr(
       logger().info("bad cmp mode {}", osd_op.op.xattr.cmp_mode);
       result = -EINVAL;
     }
-    if (result == 0) {
+    if (!result) {
```
Comment (athanatos) [inline src/crimson/osd/pg_backend.cc:1018]: This is clearer as result == 0

**Comment by athanatos** [comment]
@rzarzynski Does this look ok to you?

**Pull request merged**